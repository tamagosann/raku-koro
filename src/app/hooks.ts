import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import type { RootState, AppDispatch } from './store';

// useDispatchとかuseSelectorの代わりにこっちを使う
export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
