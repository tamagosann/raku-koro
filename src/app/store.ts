import { configureStore, ThunkAction, Action } from "@reduxjs/toolkit";
import dailyInfectionRuducer from "../features/graphs/dailyInfectionSlice";
import prefectureRuducer from "../features/common/prefectureSlice";
import dailyDeadRuducer from "../features/graphs/dailyDeadSlice";
import totalCoronaRuducer from "../features/graphs/totalCoronaSlice";
import totalDethRuducer from "../features/graphs/totalDethSlice";
import vaccinationRuducer from "../features/graphs/vaccinationSlice";
import bedOccupancyRateRuducer from "../features/graphs/bedOccupancyRateSlice";
import dailyPositiveRuducer from "../features/graphs/dailyPositiveSlice";

export const store = configureStore({
  reducer: {
    dailyInfection: dailyInfectionRuducer,
    prefecture: prefectureRuducer,
    dailyDead: dailyDeadRuducer,
    totalCorona: totalCoronaRuducer,
    totalDeth: totalDethRuducer,
    vaccination: vaccinationRuducer,
    bedOccupancyRate: bedOccupancyRateRuducer,
    dailyPositive: dailyPositiveRuducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
