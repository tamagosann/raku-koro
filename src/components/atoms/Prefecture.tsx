import React, { FC } from "react";

// common.jsから都道府県の情報を取得する
import { prefectures } from "../../common/prefecture";

// マテリアルUI
import { createStyles, makeStyles, Theme } from "@material-ui/core/styles";
import InputLabel from "@material-ui/core/InputLabel";
import MenuItem from "@material-ui/core/MenuItem";
import FormHelperText from "@material-ui/core/FormHelperText";
import FormControl from "@material-ui/core/FormControl";
import Select from "@material-ui/core/Select";

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    formControl: {
      margin: theme.spacing(1),
      minWidth: 120,
      display: "block",
    },
    selectEmpty: {
      marginTop: theme.spacing(2),
    },
    contentCenter: {
      display: "block",
      width: "300px",
      margin: "0 auto",
      marginBottom: "32px",
    },
    centerText: {
      textAlign: "center",
    },
    width: {
      width: "100%",
    },
  })
);

interface Props {
  prefecture: string;
  setPrefecture: React.Dispatch<React.SetStateAction<string>>;
}

const Prefecture: FC<Props> = ({ prefecture, setPrefecture }) => {
  const classes = useStyles();
  const handleChange = (event: React.ChangeEvent<{ value: unknown }>) => {
    setPrefecture(event.target.value as string);
  };

  return (
    <>
      <FormControl
        className={classes.formControl + " " + classes.contentCenter}
      >
        <InputLabel
          data-testid="select-input"
          color="secondary"
          id="demo-simple-select-helper-label"
        >
          都道府県
        </InputLabel>
        <Select
          labelId="demo-simple-select-helper-label"
          id="demo-simple-select-helper"
          value={prefecture}
          onChange={handleChange}
          className={classes.width + " " + classes.centerText}
        >
          {prefectures.map((prefecture, index) => (
            <MenuItem key={index} value={String(prefecture.prefCode)}>
              {prefecture.prefName}
            </MenuItem>
          ))}
        </Select>
        <FormHelperText className={classes.centerText}>
          都道府県を選択してください。
        </FormHelperText>
      </FormControl>
    </>
  );
};

export default Prefecture;
