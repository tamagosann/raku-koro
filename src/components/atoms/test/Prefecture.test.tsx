import React from 'react';
import {
  render,
  cleanup,
  screen,
  fireEvent,
  waitFor,
  within,
  getByRole,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { afterEach, expect } from '@jest/globals';
import { store } from '../../../app/store';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { Prefecture } from '../Prefecture';

describe('Prefectureコンポーネント', () => {
  // デモProps
  const prefectureProps = {
    prefecture: '1',
    setPrefecture: jest.fn(),
  };

  // 各テストの直前に実行される（コンポーネントの取得）
  beforeEach(() => {
    render(
      <Provider store={store}>
        <Prefecture
          prefecture={prefectureProps.prefecture}
          setPrefecture={prefectureProps.setPrefecture}
        />
      </Provider>
    );
  });

  // 各処理が通った後にアンマウントさせる
  afterEach(() => cleanup());

  describe('jest renderer test', () => {
    test('render correctly', () => {
      const tree = renderer
        .create(
          <Provider store={store}>
            <Prefecture
              prefecture={prefectureProps.prefecture}
              setPrefecture={prefectureProps.setPrefecture}
            />
          </Provider>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('mock関数の発火テスト', () => {
    test('セレクトボックスをクリックしたらmock関数が発火するかどうか', () => {
      const inputTag = screen.getByRole('button');
      fireEvent.mouseDown(inputTag);
      const listboxFirst = within(screen.getByRole('listbox'));
      fireEvent.click(listboxFirst.getByText(/青森県/));
      fireEvent.mouseDown(inputTag);
      const listboxSecond = within(screen.getByRole('listbox'));
      fireEvent.click(listboxSecond.getByText(/神奈川県/));
      expect(prefectureProps.setPrefecture).toHaveBeenCalledTimes(2);
    });
  });
});
