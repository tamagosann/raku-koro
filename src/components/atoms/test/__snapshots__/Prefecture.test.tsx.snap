// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prefectureコンポーネント jest renderer test render correctly 1`] = `
<div
  className="MuiFormControl-root makeStyles-formControl-1 makeStyles-contentCenter-3"
>
  <label
    className="MuiFormLabel-root MuiFormLabel-colorSecondary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
    data-shrink={true}
    data-testid="select-input"
    id="demo-simple-select-helper-label"
  >
    都道府県
  </label>
  <div
    className="MuiInputBase-root MuiInput-root MuiInput-underline makeStyles-width-5 makeStyles-centerText-4 MuiInputBase-formControl MuiInput-formControl"
    onClick={[Function]}
  >
    <div
      aria-haspopup="listbox"
      aria-labelledby="demo-simple-select-helper-label demo-simple-select-helper"
      className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
      id="demo-simple-select-helper"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      role="button"
      tabIndex={0}
    >
      北海道
    </div>
    <input
      aria-hidden={true}
      className="MuiSelect-nativeInput"
      onAnimationStart={[Function]}
      onChange={[Function]}
      required={false}
      tabIndex={-1}
      value="1"
    />
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSelect-icon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
    </svg>
  </div>
  <p
    className="MuiFormHelperText-root makeStyles-centerText-4 MuiFormHelperText-filled"
  >
    都道府県を選択してください。
  </p>
</div>
`;
